### ISSUE TITLE
CLI arguments overwritten by config

### ISSUE THREAD

**Issue by rusmux** [Lightning-AI/lightning#17533]
### Bug description

train.py:

```Python
cli = pl.cli.LightningCLI(
    seed_everything_default=0,
    args=args,
    parser_kwargs={"parser_mode": "omegaconf", "default_config_files": ["configs/base.yaml"]},
    auto_configure_optimizers=False,
    run=False,
)
```

configs/base.yaml:

```YAML
trainer:
  max_epochs: 1
```

configs/train.yaml:

```YAML
trainer:
  max_epochs: 2
```

When I run:

```Bash
python -m src.scripts.train --config=configs/check.yaml
```

There are 2 epochs, i.e. configs/train.yaml overwrites configs/base.yaml as expected.

However, when I run:

```Bash
python -m src.scripts.train --config=configs/check.yaml --trainer.max_epochs=3
```

There are still 2 epochs, i.e. CLI argument does not overwrite configs/train.yaml.

But if I specify argument not provided by config, it works normally. For example:

```Bash
python -m src.scripts.train --config=configs/check.yaml --trainer.accelerator=cpu
```

It uses CPU as an accelerator, because it was not provided in config/train.yaml. The same result If I remove `max_epochs=2` from config, then CLI argument would work as expected. CLI arguments seems to overwrite default configs provided in `LightningCLI`, but struggle with other configs passed later.

### What version are you seeing the problem on?

v2_0

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:               None
        - available:         False
        - version:           None
* Lightning:
        - lightning:         2.0.2
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.2
        - torch:             2.0.0
        - torchmetrics:      0.11.4
        - torchvision:       0.15.1
* Packages:
        - aiobotocore:       2.5.0
        - aiofiles:          22.1.0
        - aiohttp:           3.8.4
        - aiohttp-retry:     2.8.3
        - aioitertools:      0.11.0
        - aiosignal:         1.3.1
        - aiosqlite:         0.19.0
        - albumentations:    1.3.0
        - amqp:              5.1.1
        - antlr4-python3-runtime: 4.9.3
        - anyio:             3.6.2
        - appdirs:           1.4.4
        - appnope:           0.1.3
        - argcomplete:       3.0.8
        - argon2-cffi:       21.3.0
        - argon2-cffi-bindings: 21.2.0
        - arrow:             1.2.3
        - astor:             0.8.1
        - asttokens:         2.2.1
        - async-timeout:     4.0.2
        - asyncssh:          2.13.1
        - atpublic:          3.1.1
        - attrs:             23.1.0
        - babel:             2.12.1
        - backcall:          0.2.0
        - bandit:            1.7.5
        - beautifulsoup4:    4.12.2
        - billiard:          3.6.4.0
        - bleach:            6.0.0
        - blessed:           1.20.0
        - boto3:             1.26.76
        - botocore:          1.29.76
        - brotli:            1.0.9
        - build:             0.10.0
        - cachecontrol:      0.12.11
        - cachetools:        5.3.0
        - celery:            5.2.7
        - certifi:           2022.12.7
        - cffi:              1.15.1
        - cfgv:              3.3.1
        - charset-normalizer: 3.1.0
        - clearml:           1.10.3
        - cleo:              2.0.1
        - click:             8.1.3
        - click-didyoumean:  0.3.0
        - click-plugins:     1.1.1
        - click-repl:        0.2.0
        - colorama:          0.4.6
        - coloredlogs:       15.0.1
        - comm:              0.1.3
        - configobj:         5.0.8
        - contourpy:         1.0.7
        - crashtest:         0.4.1
        - croniter:          1.3.14
        - cryptography:      40.0.2
        - custom-plotly-templates: 2.0.1
        - cycler:            0.11.0
        - dacite:            1.7.0
        - darglint:          1.8.1
        - dateutils:         0.6.12
        - debugpy:           1.6.7
        - decorator:         5.1.1
        - deepdiff:          6.3.0
        - defusedxml:        0.7.1
        - deprecated:        1.2.13
        - dictdiffer:        0.9.0
        - dill:              0.3.6
        - diskcache:         5.6.1
        - distlib:           0.3.6
        - distro:            1.8.0
        - dnspython:         2.3.0
        - docstring-parser:  0.15
        - docutils:          0.19
        - dpath:             2.1.5
        - dulwich:           0.21.3
        - dvc:               2.46.0
        - dvc-data:          0.42.3
        - dvc-http:          2.30.2
        - dvc-objects:       0.21.2
        - dvc-render:        0.4.0
        - dvc-s3:            2.21.0
        - dvc-studio-client: 0.8.0
        - dvc-task:          0.2.1
        - eradicate:         2.2.0
        - eventlet:          0.33.3
        - exceptiongroup:    1.1.1
        - executing:         1.2.0
        - fastapi:           0.86.0
        - fastjsonschema:    2.16.3
        - fiftyone:          0.20.1
        - fiftyone-brain:    0.11.0
        - fiftyone-db:       0.4.0
        - filelock:          3.12.0
        - flake8:            4.0.1
        - flake8-bandit:     3.0.0
        - flake8-broken-line: 0.5.0
        - flake8-bugbear:    22.12.6
        - flake8-commas:     2.1.0
        - flake8-comprehensions: 3.12.0
        - flake8-debugger:   4.1.2
        - flake8-docstrings: 1.7.0
        - flake8-eradicate:  1.4.0
        - flake8-isort:      4.2.0
        - flake8-polyfill:   1.0.2
        - flake8-quotes:     3.3.2
        - flake8-rst-docstrings: 0.2.7
        - flake8-string-format: 0.3.0
        - flatten-dict:      0.4.2
        - flufl.lock:        7.1.1
        - fonttools:         4.39.3
        - fqdn:              1.5.1
        - frozenlist:        1.3.3
        - fsspec:            2023.4.0
        - ftfy:              6.1.1
        - funcy:             2.0
        - furl:              2.1.3
        - future:            0.18.3
        - gitdb:             4.0.10
        - gitpython:         3.1.31
        - glob2:             0.7
        - grandalf:          0.8
        - graphql-core:      3.2.3
        - greenlet:          2.0.2
        - h11:               0.14.0
        - h2:                4.1.0
        - hpack:             4.0.0
        - html5lib:          1.1
        - httpcore:          0.17.0
        - httpx:             0.24.0
        - huggingface-hub:   0.14.1
        - humanfriendly:     10.0
        - hydra-core:        1.3.2
        - hypercorn:         0.14.3
        - hyperframe:        6.0.1
        - identify:          2.5.23
        - idna:              3.4
        - imageio:           2.28.0
        - importlib-metadata: 6.6.0
        - importlib-resources: 5.12.0
        - inflate64:         0.3.1
        - iniconfig:         2.0.0
        - inquirer:          3.1.3
        - installer:         0.7.0
        - ipykernel:         6.22.0
        - ipython:           8.12.1
        - ipython-genutils:  0.2.0
        - ipywidgets:        8.0.6
        - isoduration:       20.11.0
        - isort:             5.12.0
        - iterative-telemetry: 0.0.8
        - itsdangerous:      2.1.2
        - jaraco.classes:    3.2.3
        - jedi:              0.18.2
        - jinja2:            3.1.2
        - jmespath:          1.0.1
        - joblib:            1.2.0
        - json5:             0.9.11
        - jsonargparse:      4.21.0
        - jsonlines:         3.1.0
        - jsonpointer:       2.3
        - jsonschema:        4.17.3
        - jupyter-client:    8.2.0
        - jupyter-contrib-core: 0.4.2
        - jupyter-contrib-nbextensions: 0.7.0
        - jupyter-core:      5.3.0
        - jupyter-events:    0.6.3
        - jupyter-highlight-selected-word: 0.2.0
        - jupyter-nbextensions-configurator: 0.6.1
        - jupyter-server:    2.5.0
        - jupyter-server-fileid: 0.9.0
        - jupyter-server-terminals: 0.4.4
        - jupyter-server-ydoc: 0.8.0
        - jupyter-ydoc:      0.2.4
        - jupyterlab:        3.6.3
        - jupyterlab-pygments: 0.2.2
        - jupyterlab-server: 2.22.1
        - jupyterlab-widgets: 3.0.7
        - kaleido:           0.2.1
        - keyring:           23.13.1
        - kiwisolver:        1.4.4
        - kombu:             5.2.4
        - lazy-loader:       0.2
        - lightning:         2.0.2
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - lockfile:          0.12.2
        - lxml:              4.9.2
        - markdown-it-py:    2.2.0
        - markupsafe:        2.1.2
        - matplotlib:        3.7.1
        - matplotlib-inline: 0.1.6
        - mccabe:            0.6.1
        - mdurl:             0.1.2
        - mistune:           2.0.5
        - mongoengine:       0.24.2
        - more-itertools:    9.1.0
        - motor:             3.1.2
        - mpmath:            1.3.0
        - msgpack:           1.0.5
        - multidict:         6.0.4
        - multivolumefile:   0.2.3
        - mypy:              1.2.0
        - mypy-extensions:   1.0.0
        - nanotime:          0.5.2
        - nbclassic:         0.5.6
        - nbclient:          0.7.4
        - nbconvert:         7.3.1
        - nbformat:          5.8.0
        - nest-asyncio:      1.5.6
        - networkx:          3.1
        - nodeenv:           1.7.0
        - notebook:          6.5.4
        - notebook-shim:     0.2.3
        - numpy:             1.24.3
        - omegaconf:         2.3.0
        - onnx:              1.13.1
        - opencv-python-headless: 4.7.0.72
        - ordered-set:       4.1.0
        - orderedmultidict:  1.0.1
        - orjson:            3.8.11
        - packaging:         23.1
        - pandas:            2.0.1
        - pandas-stubs:      2.0.0.230412
        - pandocfilters:     1.5.0
        - parso:             0.8.3
        - pathlib2:          2.3.7.post1
        - pathspec:          0.11.1
        - pbr:               5.11.1
        - pep8-naming:       0.13.2
        - pexpect:           4.8.0
        - pickleshare:       0.7.5
        - pillow:            9.5.0
        - pip:               23.1.2
        - pkginfo:           1.9.6
        - platformdirs:      2.6.2
        - plotly:            5.14.1
        - pluggy:            1.0.0
        - poetry:            1.4.2
        - poetry-core:       1.5.2
        - poetry-plugin-export: 1.3.1
        - pprintpp:          0.4.0
        - pre-commit:        3.2.2
        - priority:          2.0.0
        - prometheus-client: 0.16.0
        - prompt-toolkit:    3.0.38
        - protobuf:          3.20.3
        - psutil:            5.9.5
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - py7zr:             0.20.5
        - pybcj:             1.0.1
        - pycodestyle:       2.8.0
        - pycparser:         2.21
        - pycryptodomex:     3.17
        - pydantic:          1.10.7
        - pydocstyle:        6.3.0
        - pydot:             1.4.2
        - pyflakes:          2.4.0
        - pygit2:            1.12.0
        - pygments:          2.15.1
        - pygtrie:           2.5.0
        - pyjwt:             2.4.0
        - pymongo:           4.3.3
        - pyparsing:         3.0.9
        - pyppmd:            1.0.0
        - pyproject-hooks:   1.0.0
        - pyrsistent:        0.19.3
        - pytest:            7.3.1
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-json-logger: 2.0.7
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.2
        - pytz:              2023.3
        - pytz-deprecation-shim: 0.1.0.post0
        - pywavelets:        1.4.1
        - pyyaml:            6.0
        - pyzmq:             25.0.2
        - pyzstd:            0.15.7
        - qudida:            0.0.4
        - rapidfuzz:         2.15.1
        - rarfile:           4.0
        - readchar:          4.0.5
        - regex:             2023.3.23
        - requests:          2.29.0
        - requests-toolbelt: 0.10.1
        - restructuredtext-lint: 1.4.0
        - retrying:          1.3.4
        - rfc3339-validator: 0.1.4
        - rfc3986-validator: 0.1.1
        - rich:              13.3.5
        - ruamel.yaml:       0.17.21
        - ruamel.yaml.clib:  0.2.7
        - s3fs:              2023.4.0
        - s3transfer:        0.6.0
        - scikit-image:      0.20.0
        - scikit-learn:      1.2.2
        - scipy:             1.10.1
        - scmrepo:           0.2.1
        - send2trash:        1.8.2
        - setuptools:        67.7.2
        - shellingham:       1.5.0.post1
        - shortuuid:         1.0.11
        - shtab:             1.6.1
        - six:               1.16.0
        - smmap:             5.0.0
        - sniffio:           1.3.0
        - snowballstemmer:   2.2.0
        - sortedcontainers:  2.4.0
        - soupsieve:         2.4.1
        - sqltrie:           0.3.1
        - sse-starlette:     0.10.3
        - sseclient-py:      1.7.2
        - stack-data:        0.6.2
        - starlette:         0.20.4
        - starsessions:      1.3.0
        - stevedore:         5.0.0
        - strawberry-graphql: 0.138.1
        - sympy:             1.11.1
        - tabulate:          0.9.0
        - tenacity:          8.2.2
        - tensorboardx:      2.6
        - terminado:         0.17.1
        - texttable:         1.6.7
        - threadpoolctl:     3.1.0
        - tifffile:          2023.4.12
        - timm:              0.6.13
        - tinycss2:          1.2.1
        - tokenize-rt:       5.0.0
        - toml:              0.10.2
        - tomli:             2.0.1
        - tomlkit:           0.11.8
        - torch:             2.0.0
        - torchmetrics:      0.11.4
        - torchvision:       0.15.1
        - tornado:           6.3.1
        - tqdm:              4.65.0
        - traitlets:         5.9.0
        - trove-classifiers: 2023.4.29
        - typer:             0.7.0
        - types-decorator:   5.1.8.3
        - types-openpyxl:    3.1.0.4
        - types-paramiko:    3.0.0.8
        - types-pillow:      9.5.0.2
        - types-protobuf:    4.22.0.2
        - types-python-dateutil: 2.8.19.12
        - types-pytz:        2023.3.0.0
        - types-pyyaml:      6.0.12.9
        - types-requests:    2.29.0.0
        - types-setuptools:  67.7.0.0
        - types-six:         1.16.21.8
        - types-tabulate:    0.9.0.2
        - types-urllib3:     1.26.25.11
        - typeshed-client:   2.2.0
        - typing-extensions: 4.5.0
        - tzdata:            2023.3
        - tzlocal:           4.3
        - universal-analytics-python3: 1.1.1
        - uri-template:      1.2.0
        - urllib3:           1.26.15
        - uvicorn:           0.22.0
        - vine:              5.0.0
        - virtualenv:        20.16.5
        - voluptuous:        0.13.1
        - voxel51-eta:       0.9.0
        - vulture:           2.7
        - wcwidth:           0.2.6
        - webcolors:         1.13
        - webencodings:      0.5.1
        - websocket-client:  1.5.1
        - websockets:        11.0.2
        - wemake-python-styleguide: 0.17.0
        - wheel:             0.40.0
        - widgetsnbextension: 4.0.7
        - wrapt:             1.15.0
        - wsproto:           1.2.0
        - xattr:             0.10.1
        - xmltodict:         0.13.0
        - y-py:              0.5.9
        - yarl:              1.9.2
        - yesqa:             1.4.0
        - ypy-websocket:     0.8.2
        - zc.lockfile:       3.0.post1
        - zipp:              3.15.0
* System:
        - OS:                Darwin
        - architecture:
                - 64bit
                - 
        - processor:         arm
        - python:            3.10.10
        - version:           Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:41 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T8103

</details>

cc @carmocca @mauvilsa

**Comment by rusmux** [comment]
I found that it depends on the order of the arguments. If I swap the config and the argument, everything works.

**Comment by mauvilsa** [comment]
I am unable to reproduce the behavior you described. Are you sure it isn't because of something weird you are doing? In the bug description there are a couple of inconsistencies. First, the content of two config files is shown: `configs/base.yaml` and `configs/train.yaml`, however, the executed commands reference a third `configs/check.yaml`. And second, `args` is provided being unclear what this is doing.

To try to reproduce I created the two configs and a `train.py` as:

```python
#!/usr/bin/env python3

import lightning.pytorch as pl
import lightning.pytorch.cli

cli = pl.cli.LightningCLI(
    seed_everything_default=0,
    #args=args,
    parser_kwargs={"parser_mode": "omegaconf", "default_config_files": ["configs/base.yaml"]},
    auto_configure_optimizers=False,
    run=False,
)
```

Then I ran analogous commands replacing `check.yaml` with `train.yaml`. The results are as expected:

```shell
$ ./train.py --config=configs/train.yaml --print_config | grep max_epochs
  max_epochs: 2
$ ./train.py --config=configs/train.yaml --trainer.max_epochs=3 --print_config | grep max_epochs
  max_epochs: 3
$ ./train.py --config=configs/train.yaml --print_config | grep accelerator
  accelerator: auto
$ ./train.py --config=configs/train.yaml --trainer.accelerator=cpu --print_config | grep accelerator
  accelerator: cpu
```

**Comment by rusmux** [comment]
Sorry for inconsistence in examples and thank you for creating a minimal reproduction. I used you example, but changed the order of arguments and got the same behaviour. It boils down to order of arguments:

```Python
from lightning.pytorch.cli import LightningCLI
LightningCLI(run=False)
```

```YAML
python -m train --trainer.max_epochs=3 --config=configs/train.yaml --print_config | grep max_epochs
    max_epochs: 2
python -m train --config=configs/train.yaml --trainer.max_epochs=3 --print_config | grep max_epochs
    max_epochs: 3
```

Is this the expected behaviour? Shouldn't CLI arguments be preferred over config arguments?

**Comment by mauvilsa** [comment]
This is the expected behavior. Arguments are processed left to right in order. What is on the right will potentially override whatever was on the left. This is the standard behavior of most posix/linux command line tools. See [relation-between-lightningcli-and-argparse](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_faq.html#what-is-the-relation-between-lightningcli-and-argparse) and [override-order](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_faq.html#what-is-the-override-order-of-lightningcli).

**Comment by rusmux** [comment]
I see, thank you!
